qiskit_experiments/calibration_management/calibration_utils.py

```diff
@@ -12,51 +12,138 @@
 
 """Calibration helper functions"""
 
-from typing import List, Set
-from qiskit.pulse import ScheduleBlock, Call
+from typing import Optional, Set, Tuple
+import regex as re
+import retworkx as rx
 
+from qiskit.circuit import ParameterExpression, Parameter
+from qiskit.pulse import ScheduleBlock
 
-def used_in_calls(schedule_name: str, schedules: List[ScheduleBlock]) -> Set[str]:
-    """Find the schedules in the given list that call a given schedule by name.
+from qiskit_experiments.exceptions import CalibrationError
+
+
+def update_schedule_dependency(schedule: ScheduleBlock, dag: rx.PyDiGraph):
+    """Update a DAG of schedule dependencies.
+
+    Args:
+        schedule: A ScheduleBlock that potential has references to other schedules
+            that are already present in the dag.
+        dag: A directed acyclic graph that encodes schedule dependencies using references.
+    """
+    parent_idx = dag.add_node(schedule.name)
+    for reference in schedule.references:
+        dag.add_edge(parent_idx, dag.nodes().index(reference[0]), None)
+
+
+def used_in_references(schedule_names: Set[str], dag: rx.PyDiGraph) -> Set[str]:
+    """Find all the schedules in the DAG that reference the given schedules.
 
     Args:
-        schedule_name: The name of the callee to identify.
-        schedules: A list of potential caller schedules to search.
+        schedule_names: A list of schedules to which references may exist.
+        dag: The dag that represents the dependencies between schedule references.
 
     Returns:
-        A set of schedule names that call the given schedule.
+        A set of schedules that reference the given schedules.
     """
-    caller_names = set()
+    callers = set()
+
+    for name in schedule_names:
+        callers.update(_referred_by(name, dag))
 
-    for schedule in schedules:
-        if _used_in_calls(schedule_name, schedule):
-            caller_names.add(schedule.name)
+    return callers
 
-    return caller_names
 
+def _referred_by(schedule_name: str, dag: rx.PyDiGraph) -> Set[str]:
+    """Return all the schedules that refer to this schedule by name."""
+    referred_by = set()
+    for predecessor in dag.predecessors(dag.nodes().index(schedule_name)):
+        referred_by.add(predecessor)
+        referred_by.update(_referred_by(predecessor, dag))
```

Comment:
I was misunderstanding the behavior of retworkx. `.predecessor` only returns directly connected nodes.

Reply:
Could this be replaced by `retworkx.ancestors()`?

https://qiskit.org/documentation/retworkx/apiref/retworkx.ancestors.html

Reply:
Ah this is nice.

Reply:
We came to the conclusion that this is here because `predecessor` only give you the direct predecessor of the node. Therefore the recursive structure is needed.

Reply:
@eggerdj but why not use the `ancestors` function from retworkx?